NEST-295: Orchard Core API client (#1)
Lombiq/Orchard-Core-API-Client: Readme.md

> # Lombiq API Client for Orchard Core

Reviewer: Add changes applied to the guide below to the recipe as well. Test importing the recipe on a site set up with a recipe other than the SaaS one, it should work.

Author: Tested it, works with every set-up recipe.

## About

A client library for communicating with the [Orchard Core](https://www.orchardcore.net/) web APIs. Currently, it contains an implementation for the tenant management API and a console application for testing and demonstration.

Do you want to quickly try out this project and see it in action? Check it out in our [Open-Source Orchard Core Extensions](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions) full Orchard Core solution and also see our other useful Orchard Core-related open-source projects!

## Documentation

This project is about creating and setting up tenants through an API Client using [RestEase](https://github.com/canton7/RestEase).
The project requires the OpenId features to be enabled and set up, for easy use, enable the Deployment feature and import this [recipe](Lombiq.OrchardCoreApiClient.Tester/Recipes/Lombiq.OrchardCoreApiClient.Tester.OpenId.recipe.json) as a deployment package.

> If you want to set up the OpenId features without the recipe, you need to enable these features:

Reviewer: What about Tenants?

- OpenID Authorization Server
- OpenID Core Components
- OpenID Management Interface
- OpenID Token Validation
On the Admin Dashboard open Security -> OpenID Connect -> Settings -> Authorization server, and make sure that the "Enable Token Endpoint" and "Allow Client Credentials Flow" are enabled, and the Token Format should be JSON Web Token.
Head over to Security -> OpenID Connect -> Settings -> Token validation, and make sure the Authorization server tenant is set to Default.

> Last, but not least go to Security -> OpenID Connect -> Management -> Applications. Here you can add your Applications and you may specify your Client Id and Client Secret. The application will use these parameters for authentication since the application uses Client Credentials Flow. The Allow Client Credentials Flow option should be enabled.

Reviewer: What about enabling the "Allow Refresh Token Flow"? Also, no roles needed?

Author: We don't get any refresh tokens with the Client Credentials grant, so the Refresh Token Flow has no benefit for us. (to use the Refresh Token Flow, you need to send refresh tokens to the endpoint) 
Regarding the roles, you are right, I missed that.

> For testing, run your Orchard Core app first, and edit the [Program.cs](Lombiq.OrchardCoreApiClient.Tester/Program.cs) accordingly to your OpenID Application(Client Id, Client Secret and DefaultTenantUri) and then run the console application.

Reviewer: ```suggestion
The project requires the OpenId features to be enabled and set up. For easy use, enable the Deployment feature and import this [recipe](Lombiq.OrchardCoreApiClient.Tester/Recipes/Lombiq.OrchardCoreApiClient.Tester.OpenId.recipe.json) as a deployment package.

If you want to set up the OpenId features without the recipe, you need to enable these features:

- OpenID Authorization Server
- OpenID Core Components
- OpenID Management Interface
- OpenID Token Validation

1. On the Admin Dashboard open Security → OpenID Connect → Settings → Authorization server, and make sure that the "Enable Token Endpoint" and "Allow Client Credentials Flow" are enabled, and the Token Format should be JSON Web Token.
2. Head over to Security → OpenID Connect → Settings → Token validation, and make sure the Authorization server tenant is set to Default.
3. Last, but not least, go to Security → OpenID Connect → Management → Applications. Here you can add your applications and you may specify your Client Id and Client Secret. The application will use these parameters for authentication since the application uses Client Credentials Flow. The Allow Client Credentials Flow option should be enabled.

For testing, run your Orchard Core app first, then `Lombiq.OrchardCoreApiClient.Tester`. You may need to edit the [_Program.cs_](Lombiq.OrchardCoreApiClient.Tester/Program.cs) according to your OpenID Application's settings (`ClientId`, `ClientSecret`, and `DefaultTenantUri`) and then run the console application.
```

## Recipes

> Lombiq OrchardCore API Client OpenId - Recipe for enabling OpenId features and setting it up for the console tester. To use this recipe, enable the Deployment feature and import this as a deployment package.

Reviewer: ```suggestion
Lombiq OrchardCore API Client OpenId - Recipe for enabling OpenId features and setting it up for the console tester app. To use this recipe, enable the Deployment feature and import this as a deployment package.
```

> # Lombiq OrchardCore API Client

Reviewer: Use this in the other Readme and the recipe too.

```suggestion
# Lombiq API Client for Orchard Core
```

Reviewer: As said, use this in the other Readme and the recipe too.

## About

> This project contains a library with an implementation for an API communicating with the tenants and a console application for testing and demonstrating.

Reviewer: ```suggestion
A client library for communicating with the [Orchard Core](https://www.orchardcore.net/) web APIs. Currently, it contains an implementation for the tenant management API and a console application for testing and demonstration.
```

Do you want to quickly try out this project and see it in action? Check it out in our [Open-Source Orchard Core Extensions](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions) full Orchard Core solution and also see our other useful Orchard Core-related open-source projects!

## Documentation

This project is about creating and setting up tenants through an API Client using [RestEase](https://github.com/canton7/RestEase).

> The project requires the OpenId features to be enabled and set up, for easy use, enable the deployment feature and import this [recipe](Lombiq.OrchardCoreApiClient.Tester/Recipes/Lombiq.OrchardCoreApiClient.Tester.OpenId.recipe.json) as a deployment package.

Reviewer: Describe here, step by step, how you'd use the client, both for testing it out and both in another app. Also, what do you need to do in your Orchard app to make this work?

## Recipes

> Lombiq OrchardCore API Client OpenId - Recipe for enabling OpenId features and setting it up for the console tester.

Reviewer: Explain how you'd use this.

## Documentation

This project is about creating and setting up tenants through an API Client using [RestEase](https://github.com/canton7/RestEase).

> The project requires the Tenants feature and OpenId features to be enabled and set up. For easy use, enable the Deployment feature and import this [recipe](Lombiq.OrchardCoreApiClient.Tester/Recipes/Lombiq.OrchardCoreApiClient.Tester.OpenId.recipe.json) as a deployment package.

Reviewer: ```suggestion
The project requires the Tenants feature and OpenId features to be enabled and set up. For easy use, enable the Deployment feature and import [this recipe](Lombiq.OrchardCoreApiClient.Tester/Recipes/Lombiq.OrchardCoreApiClient.Tester.OpenId.recipe.json) as a deployment package.
```
